Fix connection linger for WebSTOMP

csicar opened the pull request:
## Proposed Changes

Please describe the big picture of your changes here to communicate to the RabbitMQ team why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue.

A pull request that doesn't explain **why** the change was made has a much lower chance of being accepted.

If English isn't your first language, don't worry about it and try to communicate the problem you are trying to solve to the best of your abilities.
As long as we can understand the intent, it's all good.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the `CONTRIBUTING.md` document
- [ ] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] All tests pass locally with my changes
- [ ] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [ ] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

## Further Comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc.

MarcialRosales committed:
Support Idp initiated logon in mgt ui with Oauth

Configure preferred username from a token
Make client_secret optional

MarcialRosales committed:
Support Idp initiated logon in mgt ui with Oauth

Configure preferred username from a token
Make client_secret optional

MarcialRosales committed:
Support path_prefix in /login url

For OAuth2 idp initiated logon

MarcialRosales committed:
Support path_prefix in /login url

For OAuth2 idp initiated logon

MarcialRosales committed:
Improve coverage

MarcialRosales committed:
Improve coverage

MarcialRosales committed:
Test basic-auth behind proxy

MarcialRosales committed:
Test basic-auth behind proxy

MarcialRosales committed:
Do not use cookies to store basic auth

Instead use localStorage

MarcialRosales committed:
Do not use cookies to store basic auth

Instead use localStorage

MarcialRosales committed:
Do not use cookies to store basic auth

Use window.localStorage. If it is disabled
Management UI should stop working rather than
falling back to cookies

MarcialRosales committed:
Do not use cookies to store basic auth

Use window.localStorage. If it is disabled
Management UI should stop working rather than
falling back to cookies

ansd committed:
Make classic queues v2 memory efficient

Store directory names as binary instead of string.

This commit saves >1GB of memory per 100,000 classic queues v2.
With longish node names, the memory savings are even much higher.

This commit is especially a prerequisite for scalalbe MQTT where every
subscribing MQTT connection creates its own classic queue.

So, with 3 million MQTT subscribers, this commit saves >30 GB of memory.

This commits stores file names as binaries and converts back to
file:filename() when passed to file API functions.
This is to reduce risk of breaking behaviour for path names containing
unicode chars on certain platforms.

Alternatives to the implementation in this commit:
1. Store common directory list prefix only once (e.g. put it into
   persistent_term) and store per queue directory names in ETS.
2. Use file:filename_all() instead of file:filename() and pass binaries
   to the file module functions. However this might be brittle on some
   platforms since these binaries are interpreted as "raw filenames".
   Using raw filenames requires more changes to classic queues which we
   want to avoid to reduce risk.

The downside of the implemenation in this commit is that the binary gets
converted to a list sometimes.
This happens whenever a file is flushed or a new file gets created for
example.

Following perf tests did not show any regression in performance:
```
java -jar target/perf-test.jar -s 10 -x 1 -y 0 -u q -f persistent -z 30
java -jar target/perf-test.jar -s 10000 -x 1 -y 0 -u q -f persistent -z 30
java -jar target/perf-test.jar -s 10 -x 100 -qp q%d -qpf 1 -qpt 100 -y 0 -f persistent -z 60 -c 1000
```

Furthermore `rabbit_file` did not show up in the CPU flame graphs
either.

dumbbell committed:
rabbit_feature_flags: Support virgin node init with feature flags v2

So far, when a new node is started, the feature flags v1 code is used to
enable all stable feature flags (the default behavior). In the list,
there would be `feature_flags_v2`, so this one would be enabled first
and all others would be using the v2 code path.

If we mark `feature_flags_v2` as required, the v2 code path didn't
handle this so far. This patch adds this missing code. Now, a virgin
node's feature flags can be enabled fully using the v2 code path.

As part of that, th `get_forced_feature_flag_names()` function is moved
from `rabbit_feature_flags` to `rabbit_ff_controller`.

MarcialRosales committed:
Test with selenium export/import definitions

MarcialRosales committed:
Test with selenium export/import definitions

MarcialRosales committed:
Use only token in authorization header to

import and export definitions

MarcialRosales committed:
Use only token in authorization header to

import and export definitions

dumbbell committed:
rabbit_feature_flags: Mark `feature_flags_v2` as required

This means users of RabbitMQ 3.10.x or older will have to upgrade to
RabbitMQ 3.11.x and enable `feature_flags_v2` before they can upgrade to
a more recent release.

MarcialRosales committed:
Fix import/export test cases

When running test remotely, the import
definition file must be transfered from
the test container onto the selenium container
wehre the browser runs

MarcialRosales committed:
Fix import/export test cases

When running test remotely, the import
definition file must be transfered from
the test container onto the selenium container
wehre the browser runs

ansd committed:
Test Unicode char in queue names

MarcialRosales committed:
Rename section otherwise

clicking on any area triggered
the download of the definitions

MarcialRosales committed:
Rename section otherwise

clicking on any area triggered
the download of the definitions

MarcialRosales committed:
Refactor code following recommendations

from @essen

MarcialRosales committed:
Refactor code following recommendations

from @essen

dumbbell committed:
rabbit_feature_flags: Initialize registry in `rabbit_ff_registry:inventory/0`

With the way the registry is used in the feature flags controller, the
registry was never initialized if `feature_flags_v2` was enabled
out-of-the-box (i.e. once marked as required).

The bug was discovered after marking `feature_flags_v2` as required and
removing the v1 code: the `feature_flags_v2_SUITE` testsuite failed
because the controller got an empty inventory when it called
`rabbit_ff_registry:inventory()`.

dumbbell committed:
rabbit_feature_flags: Make several fixes to `enable_feature_flag_when_ff_file_is_unwritable` testcase

The testcase was moved to another testsuite but called functions from
its initial testsuite. That code code relies on the name of the
testsuite so it broke.

At the same time, make it depend on the testsuite's feature flag, not
stream queues. The stream queues feature flag will become required at
some point and break the testcase.

dumbbell committed:
Bazel: Update number of shards for `feature_flags_SUITE`

dumbbell committed:
Bazel: Add `feature_flags_v2` to RABBITMQ_FEATURE_FLAGS

This is required for mixed-version testing where all nodes need to start
with the required feature flags enabled.

MarcialRosales committed:
Test token refresh

And enable hard session timeout which was
disabled by mistake when moved credentials
from cookies to local storage

MarcialRosales committed:
Test token refresh

And enable hard session timeout which was
disabled by mistake when moved credentials
from cookies to local storage

MarcialRosales committed:
Validate imported user from file

MarcialRosales committed:
Validate imported user from file

MarcialRosales committed:
Run selenium GH Action on oidc_idp_initiated_login branch

This is to force GH Action run selenium tests.
for some reason it is not running it

MarcialRosales committed:
Run selenium GH Action on oidc_idp_initiated_login branch

This is to force GH Action run selenium tests.
for some reason it is not running it

MarcialRosales committed:
Remove branch

MarcialRosales committed:
Remove branch

MarcialRosales committed:
Resolve conflict

MarcialRosales committed:
Resolve conflict

csicar committed:
Fix connection linger for WebSTOMP

michaelklishin committed:
Cosmetics

(cherry picked from commit 042725d8364bac3fed40df4dcdb534728dd56576)

michaelklishin committed:
Cosmetics

(cherry picked from commit 042725d8364bac3fed40df4dcdb534728dd56576)

schlagert commented on deps/rabbitmq_web_stomp/src/rabbit_web_stomp_handler.erl:
```diff
@@ -227,8 +227,12 @@ websocket_info({'EXIT', From, Reason},
         stop(State#state{ proc_state = ProcState });
     unknown_exit ->
-        stop(State)
+        self() ! close_websocket, % Allow the server to send remaining error messages
```
Use comments with double `%%` as some formatters (e.g. Emacs) format single percent comments differently.

iliakur committed:
DRY: Link docs for `/metrics/detailed` endpoint to website.

schlagert commented on deps/rabbitmq_web_stomp/src/rabbit_web_stomp_handler.erl:
```diff
@@ -291,7 +295,8 @@ handle_data1(Bytes, State = #state{proc_state  = ProcState,
                 {stop, _Reason, ProcState1} ->
-                    stop(State#state{ proc_state = ProcState1 })
+                    self() ! close_websocket,
+                    {ok, State#state{ proc_state = ProcState1 }}
```
Add comment why not stopping here directly:
`%%  do not exit here immediately, because we need to wait for messages eventually enqueued by process_request`

michaelklishin committed:
Merge pull request #6015 from rabbitmq/oidc_idp_initiated_login

OAuth 2.0: support IDP-initiated login in the management UI

michaelklishin committed:
Wording

michaelklishin committed:
Merge branch 'iliakur-ik/link-detailed-endpoint-docs-to-website'

acogoluegnes committed:
Format stream plugin code

acogoluegnes committed:
Filter out nodes in maintenance in stream metadata

We don't want clients to try to connect to those.

Fixes #3370

acogoluegnes committed:
Remove console output

dumbbell committed:
rabbit_feature_flags: Sync enabled feature flags differently on virgin node

If a virgin node starts as clustered (thanks to peer discovery), we need
to mark feature flags already enabled remotely as enabled locally too.

We can't do a regular cluster sync because remote nodes may have
required feature flags which are disabled locally on the virgin node.
Therefore, those nodes don't have the migration code for these feature
flags anymore and the feature flags' state can't be changed.

By doing this special sync, we allow a clustered virgin node to join a
cluster with remote nodes having required feature flags.

michaelklishin committed:
Merge pull request #6790 from rabbitmq/rabbitmq-3370-stream-metadata-maintenance-mode

Filter out nodes in maintenance in stream metadata

michaelklishin committed:
Merge pull request #6684 from rabbitmq/cqv2-memory

Reduce memory by multiple GBs with many classic queues

michaelklishin committed:
3.10.14 release notes

michaelklishin committed:
Merge pull request #6797 from rabbitmq/mk-3.10.14-release-notes

3.10.14 release notes

michaelklishin committed:
3.11.6 release notes

michaelklishin committed:
Merge pull request #6799 from rabbitmq/mk-3.11.6-release-notes

3.11.6 release notes

HoloRin committed:
format with buildifier

HoloRin committed:
Add "bazel run //tools:symlink_deps_for_erlang_ls"

for symlinking 3rd party deps into a place that erlang_ls can use them

HoloRin committed:
Make erlang_ls links a bit more stable

HoloRin committed:
Remove unused rule

HoloRin committed:
Remove 'w' option from ln call in //tools:symlink_deps_for_erlang_ls

as it's not a universal option, apparently

dumbbell committed:
Merge pull request #6791 from rabbitmq/enable-remote-ff-on-virgin-node

rabbit_feature_flags: Sync enabled feature flags differently on virgin node

johanrhodin committed:
add affiliation for Patrik (dentarg)

michaelklishin committed:
Merge pull request #6809 from johanrhodin/patch-9

dumbbell committed:
rabbit_feature_flags: Export rabbit_ff_controller functions for internal use

The two exported functions are used by the testsuite. They should be
exported when `TEST` is defined only, but the way mixed-version testing
is implemented in Bazel forces us to export them.

dumbbell committed:
rabbit_feature_flags: Remove v1 code

`feature_flags_v2` is required now, so we can drop the old code path.

dumbbell committed:
rabbit_feature_flags: Ack behavior change with plugins in feature_flags_SUITE

A plugin's stable feature flag will now be enabled on initial node
start with `feature_flags_v2` required.

dumbbell committed:
Merge pull request #6682 from rabbitmq/add-feature-flags-v2-virgin-node-init-support

Feature flags: Add virgin node init support to v2 code path

lukebakken committed:
Add load_definitions example

Pointed out by @tinakurian-pki

https://groups.google.com/g/rabbitmq-users/c/UAi_v3dyMtQ

michaelklishin committed:
Merge pull request #6814 from rabbitmq/lukebakken/add-load-defs-example

dependabot[bot] committed:
Bump actions/upload-artifact from 3.1.1 to 3.1.2

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 3.1.1 to 3.1.2.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v3.1.1...v3.1.2)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

michaelklishin committed:
Merge pull request #6815 from rabbitmq/dependabot/github_actions/main/actions/upload-artifact-3.1.2

dumbbell committed:
rabbit_feature_flags: Drop the `-on_load()` directive

We used it to prevent a node breakage when RabbitMQ was updated from
e.g. 3.6.x to 3.7.0 when the package manager uncompresses the files of
the package while the service is running, overwriting existing files. In
this situation, a 3.6.x node could be queried for its feature flags
state and the `rabbit_feature_flags` module would be loaded (because it
appeared during the upgrade but before the restart) and this would fail
because the node really didn't support feature flags.

With some feature flags being required as of RabbitMQ 3.11.0, upgrading
directly from e.g. 3.6.x is unsupported now. Therefore, we can drop this
`-on_load()` failsafe.

dumbbell committed:
rabbit_feature_flags: Remove pre-3.7.x error messages

They were useful to diagnose failures when upgrading from e.g. 3.6.x to
3.8.x, but this is no longer supported. Therefore, drop this part of the
code.

dependabot[bot] committed:
Bump actions/cache from 3.2.2 to 3.2.3

Bumps [actions/cache](https://github.com/actions/cache) from 3.2.2 to 3.2.3.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v3.2.2...v3.2.3)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

michaelklishin committed:
Merge pull request #6822 from rabbitmq/dependabot/github_actions/main/actions/cache-3.2.3

csicar committed:
Fix tests: Do not close connection from client side

- Server closes connection on the error case, so the client was not able
  to do so and consequently hung
- Do not test for the error code returned by the server, because it is
  not part of the spec and now accessible via the websocket client,
  which handles the TCP close instead of the Websocket close message

HoloRin committed:
Merge pull request #6807 from rabbitmq/bazel-erlang_ls

erlang_ls workaround for bazel

dumbbell committed:
rabbit_feature_flags: Fix {enable,post_enable}_callback() type spec

The type declared the callback function would take 3 arguments instead
of 1. The single argument is a map which contains the feature flag name
and properties.

dumbbell committed:
rabbit_feature_flags: Remove incorrect paragraph in `enable_callback()` doc

dumbbell committed:
Merge pull request #6831 from rabbitmq/feature-flags-type-spec-and-doc-fixes

Feature flags: Fixes to type specs and doc

dumbbell committed:
rabbit_feature_flags: Remove support for v1 migration functions

All existing feature flags either use v2 callbacks or don't have any
associated migration code. Future flags will have to use v2 callbacks if
they need to.

v1 migration functions were deprecated with the introduction of the
feature flags controller in RabbitMQ 3.11.0. Removing support for v1
migration functions simplifies the code.

dumbbell committed:
rabbit_feature_flags: Use `?LOG_*()` macros for logging

... instead of the internal `rabbit_log_feature_flags` module.

dumbbell committed:
Merge pull request #6810 from rabbitmq/make-feature_flags_v2-required

Feature flags: Make feature flags v2 required
